Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift

```diff
@@ -0,0 +1,221 @@
+
+      let note = Note(
+        node: noteNode,
+        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
+      )
+      notes.append(note)
+    }
+
+    return Diagnostic(
+      node: diagnosticNode,
+      message: SimpleDiagnosticMessage(
+        message: self.message,
+        diagnosticID: self.id,
+        severity: self.severity
+      ),
+      highlights: self.highlight,
+      notes: notes,
+      fixIts: self.fixIts
+    )
+  }
+}
+
+/// Represents a descriptor for constructing a note message in testing.
+struct NoteDescriptor {
+  /// The marker pointing to location in source code.
+  let locationMarker: LocationMarker
+
+  /// The ID associated with the note message.
+  let id: MessageID
+
+  /// The textual content of the note to be displayed.
+  let message: String
+}
+
+/// A simple implementation of the `NoteMessage` protocol for testing.
+/// This struct holds the message text and a fix-it ID for a note.
+struct SimpleNoteMessage: NoteMessage {
+  /// The textual content of the note to be displayed.
+  let message: String
+
+  /// The ID associated with the fix-it, if applicable.
+  let fixItID: MessageID
+}
+
+/// A simple implementation of the `DiagnosticMessage` protocol for testing.
+/// This struct holds the message text, diagnostic ID, and severity for a diagnostic.
+struct SimpleDiagnosticMessage: DiagnosticMessage {
+  /// The textual content of the diagnostic message to be displayed.
+  let message: String
+
+  /// The ID associated with the diagnostic message for categorization or referencing.
+  let diagnosticID: MessageID
+
+  /// The severity level of the diagnostic message.
+  let severity: DiagnosticSeverity
+}
+
+/// Asserts that the annotated source generated from diagnostics matches an expected annotated source.
+///
+/// - Parameters:
+///   - markedSource: The source code with location markers `LocationMarker` for diagnostics.
+///   - withDiagnostics: An array of diagnostic descriptors to generate diagnostics.
+///   - matches: The expected annotated source after applying the diagnostics.
+///   - file: The file in which failure occurred.
+///   - line: The line number on which failure occurred.
+func assertAnnotated(
+  markedSource: String,
+  withDiagnostics diagnosticDescriptors: [DiagnosticDescriptor],
+  matches expectedAnnotatedSource: String,
+  file: StaticString = #file,
+  line: UInt = #line
+) {
+
+  let (markers, source) = extractMarkers(markedSource)
+  let tree = Parser.parse(source: source)
+
+  // Check for any unexpected diagnostics generated by parser.
+  let generatedDiagnostics = ParseDiagnosticsGenerator.diagnostics(for: tree)
+  if !generatedDiagnostics.isEmpty {
+    let diagnosticMessages = generatedDiagnostics.map { $0.message }.joined(separator: ", ")
+    XCTFail("Unexpected diagnostics found in the source: \(diagnosticMessages)", file: file, line: line)
+  }
```

Comment:
Why do we need to check this? I think it would be totally fine to add our own diagnostics based on the `DiagnosticDescriptor`s even if the syntax tree contains errors.

Reply:
Sounds good to me!

Reply:
Removed